### ISSUE TITLE
Parallelization of the generate command not using available CPU resource 

### ISSUE THREAD

**Issue by millionpoundhat** [t-rex-tileserver/t-rex#162]
Hello, 

I'm using the latest docker container --version
t_rex 0.10.0 (GDAL version 2.2.2). When I run the web server, request tiles on the fly and monitor the CPU usage t-rex is using all the available cores at around 100% which is expected as per the documentation. 

When I run a generate command without setting (--nodes or --nodeno) I would expect similar behavior but this doesn't happen. Do I need to configure --nodes or --nodeno to force parallelization? I've tried various combinations relating to my system and not managed to get the same performance from the generate command. 

Please advise. Thanks in advance.

**Comment by pka** [comment]
t-rex doesn't parallelize automatically yet, so you have to add `--nodes`and `--nodeno` to use multiple cores, processors or machines.

**Comment by millionpoundhat** [comment]
Thanks, 

So what do the --nodes and --nodeno flags do specifically?

**Comment by pka** [comment]
--nodes <NUM>                               Number of generator nodes
        --nodeno <NUM>                              Number of this nodes (0 <= n < nodes)

See https://t-rex.tileserver.ch/doc/generate/ for all options.

**Comment by millionpoundhat** [comment]
Thanks pka, 

I've read the documentation. But it is not clear (to me, sorry) which of these controls is setting those parameters. If I have 1 machine with 4 cores, would I set --nodes 1 and --nodeno 4?

**Comment by pka** [comment]
Then you start t_rex 4 times in parallel:
```
t_rex generate --nodes 4 --nodeno 0 &
t_rex generate --nodes 4 --nodeno 1 &
t_rex generate --nodes 4 --nodeno 2 &
t_rex generate --nodes 4 --nodeno 3
` ``

**Comment by millionpoundhat** [comment]
That's great thanks for the explanation @pka

**Comment by stvno** [comment]
Still it would be nice if it would automatically do so when setting the number of nodes, running 10 nodes with an extent becomes rather cumbersome this way: 
```
t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 0 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,2259
8.080,595401.92,903401.92 --nodes 10 --nodeno 1 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 2 & t_rex generate --config cartiqo-rd.toml --exten
t -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 3 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 4 & t_rex generate --config cartiq
o-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 5 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 6 & t_rex generat
e --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 7 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno
8 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 9 
```

**Comment by olefemmy** [comment]
> Still it would be nice if it would automatically do so when setting the number of nodes, running 10 nodes with an extent becomes rather cumbersome this way:
> 
> ```
> t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 0 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,2259
> 8.080,595401.92,903401.92 --nodes 10 --nodeno 1 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 2 & t_rex generate --config cartiqo-rd.toml --exten
> t -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 3 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 4 & t_rex generate --config cartiq
> o-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 5 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 6 & t_rex generat
> e --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 7 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno
> 8 & t_rex generate --config cartiqo-rd.toml --extent -285401.92,22598.080,595401.92,903401.92 --nodes 10 --nodeno 9 
> ```

You can run it like this?

`gv_Core=$(nproc --all)
#gv_Core=$(getconf _NPROCESSORS_ONLN)

gv_Minzoom=${1:-6}
gv_Maxzoom=${2:-19}

# Run a loop for generate

for ii in $(seq 0 ${gv_Core})
do
    /usr/bin/t_rex generate --config cartiqo-rd.toml --minzoom ${gv_Minzoom} --maxzoom ${gv_Maxzoom} --nodes ${gv_Core} --nodeno $ii &
done
`

**Comment by pka** [comment]
Since async/await landed in Rust stable a few days ago, writing a async implementation is now a promising option.

@stvno: the current command line options are necessary for running parallel jobs on different machines, because automatic parallelization over machines is different story.

I'm relabelling this issue as a reminder for automatic paralelization on a single machine though.

**Comment by cristianmad** [comment]
> Since async/await landed in Rust stable a few days ago, writing a async implementation is now a promising option.
> 
> @stvno: the current command line options are necessary for running parallel jobs on different machines, because automatic parallelization over machines is different story.
> 
> I'm relabelling this issue as a reminder for automatic paralelization on a single machine though.

Any progress on this? :) 

I'm trying the approach indicated in this thread and no matter how many nodes I create (4 in my case), I do not believe any parallelization is done.
For example, this is the command in run (Windows 10)

_t_rex.exe generate --config D:\myconfig.toml --extent -171.791110603,18.91619,-66.96466,71.3577635769 --nodes 4 --nodeno 0 & t_rex.exe generate --config D:\myconfig.toml --extent -171.791110603,18.91619,-66.96466,71.3577635769 --nodes 4 --nodeno 1 & t_rex.exe generate --config D:\myconfig.toml --extent -171.791110603,18.91619,-66.96466,71.3577635769 --nodes 4 --nodeno 2 & t_rex.exe generate --config D:\myconfig.toml --extent -171.791110603,18.91619,-66.96466,71.3577635769 --nodes 4 --nodeno 3_

I can only see a single PostgrSQL process. However, if I serve instead of trying to generate, I see multiple instances running and my CPU being fully used (as desired).

Any help would be appreciated.
Thank you!

**Comment by olefemmy** [comment]
I'm not sure that's the way to run the command in background as it seems you're in Windows environment? See [here](https://superuser.com/questions/198525/how-can-i-execute-a-windows-command-line-in-background) for more information.

**Comment by pka** [comment]
Automatic distribution on available CPU cores is implemented with #232
